File: src/amplifier_app_newtui/ui/rewind_strip.py
"""Rewind picker overlay strip (DESIGN-SPEC §9, §2 overlay strips).

A bordered orange strip docked ABOVE the composer, opened by ctrl-r /
``/rewind`` / clicking a turn rule:

``‹ rewind › tN · $X.XX · <label> › [enter fork] [esc close]``

- ``‹`` / ``›`` (click or ``←``/``→``) navigate checkpoints, clamped at
  the ends (mockup ``Math.max/Math.min`` — no wrap-around).
- ``enter fork`` (chip, bright on bg-tab; Enter or click) posts
  :class:`RewindStrip.ForkRequested` with the current checkpoint id —
  the app performs the actual session fork (confirm-then-trim,
  ADR-0007) and only then trims the transcript.
- ``esc close`` (dimmer; Esc or click) posts :class:`RewindStrip.Closed`.

The strip hides itself after fork/close.
"""

from __future__ import annotations

from collections.abc import Sequence

from textual import events
from textual.binding import Binding
from textual.containers import Horizontal
from textual.message import Message
from textual.widgets import Static

from ..model.blocks import GLYPH_REWIND_LEFT, GLYPH_REWIND_RIGHT
from ..model.turn import Checkpoint

FORK_HINT = "enter fork"
CLOSE_HINT = "esc close"


def rewind_label(checkpoint: Checkpoint) -> str:
    """``tN · $X.XX · <label>`` — the picker's checkpoint description."""
    return f"{checkpoint.id} · ${checkpoint.cost_at:.2f} · {checkpoint.label}"


def rewind_line(checkpoint: Checkpoint) -> str:
    """The strip's center text: ``rewind › tN · $X.XX · <label>``."""
    return f"rewind {GLYPH_REWIND_RIGHT} {rewind_label(checkpoint)}"


class RewindStrip(Horizontal):
    """The rewind picker strip (DESIGN-SPEC §9).

    Open with :meth:`show_checkpoints` (defaults to the newest
    checkpoint, or the clicked rule's). Posts:

    - :class:`ForkRequested` — Enter / ``enter fork`` chip click.
    - :class:`Closed` — Esc / ``esc close`` click.
    """

    can_focus = True

    DEFAULT_CSS = """
    RewindStrip {
        display: none;
        width: 100%;
        height: auto;
        border-top: solid $rule;
        padding: 0 2;
        color: $orange;
    }
    RewindStrip > Static {
        width: auto;
        height: 1;
        color: $orange;
        margin-right: 1;
    }
    RewindStrip #rewind-fork {
        color: $bright;
        background: $bg-tab;
        padding: 0 1;
    }
    RewindStrip #rewind-close {
        color: $dimmer;
    }
    """

    BINDINGS = [
        Binding("left", "prev", "‹ ›", show=False),
        Binding("right", "next", "‹ ›", show=False),
        Binding("enter", "fork", "enter fork", show=False),
        # No local escape binding: Esc must bubble to the app so it resolves
        # via keymap.ESC_CHAIN (spec §5 — lane-focus/palette close before
        # rewind even while this strip holds keyboard focus). The chain
        # calls ``action_close`` when the rewind step is reached.
    ]

    class ForkRequested(Message):
        """The user asked to fork from a checkpoint (Enter / chip click)."""

        def __init__(self, checkpoint_id: str) -> None:
            self.checkpoint_id = checkpoint_id
            super().__init__()

    class Closed(Message):
        """Esc pressed / ``esc close`` clicked."""

    class TypeThrough(Message):
        """A printable key pressed while the strip held focus.

        Mockup ground truth (document-level keydown, composer input keeps
        focus while ``rewindOpen``): typing is never swallowed by the
        rewind picker — the app forwards the character to the composer,
        so ``/`` opens the palette live-filtered and the text lands in
        the input (spec §5).
        """

        def __init__(self, character: str) -> None:
            self.character = character
            super().__init__()

    def __init__(self, *, id: str | None = None) -> None:  # noqa: A002
        super().__init__(id=id)
        self._checkpoints: tuple[Checkpoint, ...] = ()
        self._index = 0

    def compose(self):
        yield Static(GLYPH_REWIND_LEFT, id="rewind-prev")
        yield Static("", id="rewind-label")
        yield Static(GLYPH_REWIND_RIGHT, id="rewind-next")
        yield Static(FORK_HINT, id="rewind-fork")
        yield Static(CLOSE_HINT, id="rewind-close")

    # -- public API ----------------------------------------------------

    @property
    def checkpoints(self) -> tuple[Checkpoint, ...]:
        return self._checkpoints

    @property
    def index(self) -> int:
        return self._index

    @property
    def current(self) -> Checkpoint | None:
        if not self._checkpoints:
            return None
        return self._checkpoints[self._index]

    @property
    def label_text(self) -> str:
        """The exact center text currently displayed."""
        current = self.current
        return rewind_line(current) if current is not None else ""

    def show_checkpoints(
        self, checkpoints: Sequence[Checkpoint], index: int | None = None
    ) -> None:
        """Open the picker on *checkpoints* (newest selected by default).

        An empty checkpoint list keeps the strip hidden — the app shows
        the ``no rewind checkpoints yet`` notice instead.
        """
        self._checkpoints = tuple(checkpoints)
        if not self._checkpoints:
            self.display = False
            return
        last = len(self._checkpoints) - 1
        self._index = last if index is None else max(0, min(last, index))
        self._refresh_label()
        self.display = True
        self.focus()

    def sync_checkpoints(self, checkpoints: Sequence[Checkpoint]) -> None:
        """Refresh the open picker's list in place (mockup openRewind /
        rewindNext read the live ``this.checkpoints`` array — a checkpoint
        cut while the picker is open is immediately navigable with ›).

        The cursor position is preserved (clamped); focus is untouched.
        """
        if not self.display:
            return
        self._checkpoints = tuple(checkpoints)
        if not self._checkpoints:
            self.display = False
            return
        self._index = max(0, min(len(self._checkpoints) - 1, self._index))
        self._refresh_label()

    def nav(self, delta: int) -> None:
        """Move the checkpoint cursor by *delta*, clamped at both ends."""
        if not self._checkpoints:
            return
        self._index = max(0, min(len(self._checkpoints) - 1, self._index + delta))
        self._refresh_label()

    def fork(self) -> None:
        """Request the fork for the current checkpoint and close the strip."""
        current = self.current
        if current is None:
            return
        self.display = False
        self.post_message(self.ForkRequested(current.id))

    def close_strip(self) -> None:
        self.display = False
        self.post_message(self.Closed())

    # -- key actions ----------------------------------------------------

    def on_key(self, event: events.Key) -> None:
        """Printable keys pass through to the composer (mockup: the
        composer keeps typing rights while ``rewindOpen``); ←→/enter stay
        with the strip via BINDINGS, esc bubbles to the app's ESC_CHAIN."""
        if event.is_printable and event.character:
            event.stop()
            event.prevent_default()
            self.post_message(self.TypeThrough(event.character))

    def action_prev(self) -> None:
        self.nav(-1)

    def action_next(self) -> None:
        self.nav(1)

    def action_fork(self) -> None:
        self.fork()

    def action_close(self) -> None:
        self.close_strip()

    # -- clicks ----------------------------------------------------------

    def on_click(self, event: events.Click) -> None:
        widget = event.widget
        if widget is None or widget.id is None:
            return
        if widget.id == "rewind-prev":
            self.nav(-1)
        elif widget.id == "rewind-next":
            self.nav(1)
        elif widget.id == "rewind-fork":
            self.fork()
        elif widget.id == "rewind-close":
            self.close_strip()

    # -- internals -------------------------------------------------------

    def _refresh_label(self) -> None:
        if self.is_mounted:
            self.query_one("#rewind-label", Static).update(self.label_text)


__all__ = [
    "CLOSE_HINT",
    "FORK_HINT",
    "RewindStrip",
    "rewind_label",
    "rewind_line",
]

File: rust-mvp/src/ui/rewind_strip.rs
//! Rewind picker overlay strip (DESIGN-SPEC §9, §2 overlay strips) — port of
//! `ui/rewind_strip.py`.
//!
//! A bordered orange strip docked ABOVE the composer, opened by ctrl-r /
//! `/rewind` / clicking a turn rule:
//!
//! `‹ rewind · pick a turn · turn N · $X.XX · <label> › [enter fork] [esc close]`
//!
//! - `‹` / `›` (click or `←`/`→`) navigate checkpoints, clamped at the ends
//!   (mockup `Math.max/Math.min` — no wrap-around).
//! - `enter fork` (chip, bright on bg-tab; Enter or click) emits
//!   [`RewindMsg::ForkRequested`] with the current checkpoint id — the app
//!   performs the actual session fork (confirm-then-trim, ADR-0007) and only
//!   then trims the transcript.
//! - `esc close` (dimmer; Esc or click) emits [`RewindMsg::Closed`].
//!
//! The strip hides itself after fork/close.
//!
//! Textual mechanics translated for ratatui: the widget's posted `Message`
//! classes become the [`RewindMsg`] enum returned by the action/click entry
//! points; `display`/focus are plain state the app-assembly layer reads when
//! laying out and routing keys. The Textual BINDINGS (`left`/`right`/`enter`,
//! esc via the app's ESC_CHAIN) map onto [`RewindStrip::action_prev`],
//! [`RewindStrip::action_next`], [`RewindStrip::action_fork`], and
//! [`RewindStrip::action_close`].

use crate::model::blocks::{GLYPH_REWIND_LEFT, GLYPH_REWIND_RIGHT};
use crate::model::turn::Checkpoint;

pub const FORK_HINT: &str = "enter fork";
pub const CLOSE_HINT: &str = "esc close";

/// `turn N · $X.XX · <label>` — the picker's checkpoint description.
///
/// The turn is spelled out (`turn 3`, not the cryptic `t3`) so the marker
/// reads legibly (S5 discoverability: users could not tell what `t3` meant).
pub fn rewind_label(checkpoint: &Checkpoint) -> String {
    // round_dp uses banker's rounding (MidpointNearestEven), matching the
    // default decimal context Python's `f"{Decimal:.2f}"` formats with;
    // `{:.2}` then only pads.
    format!(
        "turn {} · ${:.2} · {}",
        checkpoint.turn_id,
        checkpoint.cost_at.round_dp(2),
        checkpoint.label
    )
}

/// The strip's center text: `rewind · pick a turn · turn N · $X.XX · <label>`.
///
/// The `pick a turn` phrase turns the strip into a self-explaining header: it
/// names the feature (`rewind`), states the action, then shows the currently
/// selected turn — flanked by the ‹ › nav glyphs and the `enter fork` /
/// `esc close` chips the strip composes alongside.
pub fn rewind_line(checkpoint: &Checkpoint) -> String {
    format!("rewind · pick a turn · {}", rewind_label(checkpoint))
}

/// What the Textual widget posted as `RewindStrip.ForkRequested` /
/// `.Closed` / `.TypeThrough` messages; the ratatui app-assembly layer
/// receives these as return values from the strip's entry points.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum RewindMsg {
    /// The user asked to fork from a checkpoint (Enter / chip click).
    ForkRequested { checkpoint_id: String },
    /// Esc pressed / `esc close` clicked.
    Closed,
    /// A printable key pressed while the strip held focus.
    ///
    /// Mockup ground truth (document-level keydown, composer input keeps
    /// focus while `rewindOpen`): typing is never swallowed by the rewind
    /// picker — the app forwards the character to the composer, so `/`
    /// opens the palette live-filtered and the text lands in the input
    /// (spec §5).
    TypeThrough { character: String },
}

/// The rewind picker strip (DESIGN-SPEC §9).
///
/// Open with [`RewindStrip::show_checkpoints`] (defaults to the newest
/// checkpoint, or the clicked rule's). Emits:
///
/// - [`RewindMsg::ForkRequested`] — Enter / `enter fork` chip click.
/// - [`RewindMsg::Closed`] — Esc / `esc close` click.
#[derive(Clone, Debug, Default)]
pub struct RewindStrip {
    checkpoints: Vec<Checkpoint>,
    index: usize,
    display: bool,
}

impl RewindStrip {
    pub fn new() -> Self {
        Self::default()
    }

    // -- public API ----------------------------------------------------

    pub fn checkpoints(&self) -> &[Checkpoint] {
        &self.checkpoints
    }

    pub fn index(&self) -> usize {
        self.index
    }

    /// Whether the strip is visible (Textual `display`).
    pub fn display(&self) -> bool {
        self.display
    }

    pub fn current(&self) -> Option<&Checkpoint> {
        self.checkpoints.get(self.index)
    }

    /// The exact center text currently displayed.
    pub fn label_text(&self) -> String {
        match self.current() {
            Some(current) => rewind_line(current),
            None => String::new(),
        }
    }

    /// The strip's five children in compose order, as `(id, text)` pairs —
    /// what Textual mounted as Statics (`‹`, label, `›`, chips). The
    /// app-assembly layer styles them per DEFAULT_CSS (orange strip, bright
    /// fork chip on bg-tab, dimmer close hint).
    pub fn segments(&self) -> [(&'static str, String); 5] {
        [
            ("rewind-prev", GLYPH_REWIND_LEFT.to_string()),
            ("rewind-label", self.label_text()),
            ("rewind-next", GLYPH_REWIND_RIGHT.to_string()),
            ("rewind-fork", FORK_HINT.to_string()),
            ("rewind-close", CLOSE_HINT.to_string()),
        ]
    }

    /// Open the picker on `checkpoints` (newest selected by default).
    ///
    /// An empty checkpoint list keeps the strip hidden — the app shows the
    /// `no rewind checkpoints yet` notice instead.
    pub fn show_checkpoints(&mut self, checkpoints: &[Checkpoint], index: Option<usize>) {
        self.checkpoints = checkpoints.to_vec();
        if self.checkpoints.is_empty() {
            self.display = false;
            return;
        }
        let last = self.checkpoints.len() - 1;
        self.index = match index {
            None => last,
            Some(index) => index.min(last),
        };
        self.display = true;
        // Textual `self.focus()` is app-assembly wiring in ratatui.
    }

    /// Refresh the open picker's list in place (mockup openRewind /
    /// rewindNext read the live `this.checkpoints` array — a checkpoint cut
    /// while the picker is open is immediately navigable with ›).
    ///
    /// The cursor position is preserved (clamped); focus is untouched.
    pub fn sync_checkpoints(&mut self, checkpoints: &[Checkpoint]) {
        if !self.display {
            return;
        }
        self.checkpoints = checkpoints.to_vec();
        if self.checkpoints.is_empty() {
            self.display = false;
            return;
        }
        self.index = self.index.min(self.checkpoints.len() - 1);
    }

    /// Move the checkpoint cursor by `delta`, clamped at both ends.
    pub fn nav(&mut self, delta: i64) {
        if self.checkpoints.is_empty() {
            return;
        }
        let last = (self.checkpoints.len() - 1) as i64;
        self.index = (self.index as i64 + delta).clamp(0, last) as usize;
    }

    /// Request the fork for the current checkpoint and close the strip.
    pub fn fork(&mut self) -> Option<RewindMsg> {
        let checkpoint_id = self.current()?.id.clone();
        self.display = false;
        Some(RewindMsg::ForkRequested { checkpoint_id })
    }

    pub fn close_strip(&mut self) -> RewindMsg {
        self.display = false;
        RewindMsg::Closed
    }

    // -- key actions ----------------------------------------------------

    /// Printable keys pass through to the composer (mockup: the composer
    /// keeps typing rights while `rewindOpen`); ←→/enter stay with the strip
    /// via its bindings, esc bubbles to the app's ESC_CHAIN.
    pub fn on_printable_key(&self, character: &str) -> RewindMsg {
        RewindMsg::TypeThrough {
            character: character.to_string(),
        }
    }

    pub fn action_prev(&mut self) {
        self.nav(-1);
    }

    pub fn action_next(&mut self) {
        self.nav(1);
    }

    pub fn action_fork(&mut self) -> Option<RewindMsg> {
        self.fork()
    }

    pub fn action_close(&mut self) -> RewindMsg {
        self.close_strip()
    }

    // -- clicks ----------------------------------------------------------

    /// Dispatch a click on one of the strip's children by id (the Textual
    /// `on_click` widget-id table); unknown ids are ignored.
    pub fn on_click(&mut self, widget_id: &str) -> Option<RewindMsg> {
        match widget_id {
            "rewind-prev" => {
                self.nav(-1);
                None
            }
            "rewind-next" => {
                self.nav(1);
                None
            }
            "rewind-fork" => self.fork(),
            "rewind-close" => Some(self.close_strip()),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rust_decimal::Decimal;
    use std::str::FromStr;

    fn checkpoints() -> Vec<Checkpoint> {
        vec![
            Checkpoint {
                id: "t1".to_string(),
                turn_id: 1,
                message_index: 4,
                cost_at: Decimal::from_str("0.18").unwrap(),
                label: "store refactor · shipped".to_string(),
            },
            Checkpoint {
                id: "t2".to_string(),
                turn_id: 2,
                message_index: 9,
                cost_at: Decimal::from_str("0.47").unwrap(),
                label: "auto run · shipped locally".to_string(),
            },
            Checkpoint {
                id: "t3".to_string(),
                turn_id: 3,
                message_index: 13,
                cost_at: Decimal::from_str("1.12").unwrap(),
                label: "plan ready".to_string(),
            },
        ]
    }

    // -- pure formatting -----------------------------------------------

    #[test]
    fn test_rewind_label_exact_string() {
        let cps = checkpoints();
        assert_eq!(rewind_label(&cps[2]), "turn 3 · $1.12 · plan ready");
        assert_eq!(
            rewind_line(&cps[0]),
            "rewind · pick a turn · turn 1 · $0.18 · store refactor · shipped"
        );
    }

    #[test]
    fn test_hint_strings() {
        assert_eq!(FORK_HINT, "enter fork");
        assert_eq!(CLOSE_HINT, "esc close");
    }

    // -- widget behavior -------------------------------------------------

    #[test]
    fn test_opens_on_newest_checkpoint_by_default() {
        let mut strip = RewindStrip::new();
        assert!(!strip.display());
        strip.show_checkpoints(&checkpoints(), None);
        assert!(strip.display());
        assert_eq!(strip.index(), 2);
        assert_eq!(
            strip.label_text(),
            "rewind · pick a turn · turn 3 · $1.12 · plan ready"
        );
    }

    #[test]
    fn test_opens_at_clicked_rule_checkpoint() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), Some(0));
        assert_eq!(
            strip.label_text(),
            "rewind · pick a turn · turn 1 · $0.18 · store refactor · shipped"
        );
    }

    #[test]
    fn test_arrow_navigation_is_clamped() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        strip.action_prev();
        strip.action_prev();
        assert_eq!(strip.index(), 0);
        strip.action_prev(); // clamped at the oldest
        assert_eq!(strip.index(), 0);
        // clamped at newest
        strip.action_next();
        strip.action_next();
        strip.action_next();
        strip.action_next();
        assert_eq!(strip.index(), 2);
        assert_eq!(
            strip.label_text(),
            "rewind · pick a turn · turn 3 · $1.12 · plan ready"
        );
    }

    #[test]
    fn test_enter_requests_fork_for_current_checkpoint_and_closes() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        strip.action_prev();
        let msg = strip.action_fork();
        assert_eq!(
            msg,
            Some(RewindMsg::ForkRequested {
                checkpoint_id: "t2".to_string()
            })
        );
        assert!(!strip.display());
    }

    #[test]
    fn test_close_action_posts_closed_and_hides() {
        // Esc is resolved by the app via keymap.ESC_CHAIN (spec §5) — the
        // strip has no local escape binding; the chain invokes `action_close`.
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        let msg = strip.action_close();
        assert_eq!(msg, RewindMsg::Closed);
        assert!(!strip.display());
    }

    #[test]
    fn test_click_glyphs_navigate_and_fork_chip_forks() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        assert_eq!(strip.on_click("rewind-prev"), None);
        assert_eq!(strip.index(), 1);
        assert_eq!(strip.on_click("rewind-next"), None);
        assert_eq!(strip.index(), 2);
        assert_eq!(
            strip.on_click("rewind-fork"),
            Some(RewindMsg::ForkRequested {
                checkpoint_id: "t3".to_string()
            })
        );
        assert!(!strip.display());
    }

    #[test]
    fn test_empty_checkpoints_keep_strip_hidden() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&[], None);
        assert!(!strip.display());
    }

    // -- non-pinned behavior kept faithful to the Python source ----------

    #[test]
    fn test_close_click_posts_closed() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        assert_eq!(strip.on_click("rewind-close"), Some(RewindMsg::Closed));
        assert!(!strip.display());
    }

    #[test]
    fn test_printable_key_types_through() {
        let strip = RewindStrip::new();
        assert_eq!(
            strip.on_printable_key("/"),
            RewindMsg::TypeThrough {
                character: "/".to_string()
            }
        );
    }

    #[test]
    fn test_sync_checkpoints_clamps_cursor_and_hides_on_empty() {
        let cps = checkpoints();
        let mut strip = RewindStrip::new();
        // Ignored while hidden.
        strip.sync_checkpoints(&cps);
        assert!(strip.checkpoints().is_empty());
        strip.show_checkpoints(&cps, None);
        assert_eq!(strip.index(), 2);
        strip.sync_checkpoints(&cps[..1]);
        assert!(strip.display());
        assert_eq!(strip.index(), 0);
        strip.sync_checkpoints(&[]);
        assert!(!strip.display());
    }

    #[test]
    fn test_segments_compose_order_and_glyphs() {
        let mut strip = RewindStrip::new();
        strip.show_checkpoints(&checkpoints(), None);
        let segments = strip.segments();
        assert_eq!(segments[0], ("rewind-prev", "‹".to_string()));
        assert_eq!(segments[1].0, "rewind-label");
        assert_eq!(
            segments[1].1,
            "rewind · pick a turn · turn 3 · $1.12 · plan ready"
        );
        assert_eq!(segments[2], ("rewind-next", "›".to_string()));
        assert_eq!(segments[3], ("rewind-fork", FORK_HINT.to_string()));
        assert_eq!(segments[4], ("rewind-close", CLOSE_HINT.to_string()));
    }
}
